api/envoy/config/route/v3/route_components.proto

```diff
@@ -1099,6 +1119,15 @@ message RouteAction {
   // :ref:`HTTP_DOWNSTREAM_STREAM_IDLE <envoy_v3_api_enum_value_config.overload.v3.ScaleTimersOverloadActionConfig.TimerType.HTTP_DOWNSTREAM_STREAM_IDLE>`.
   google.protobuf.Duration idle_timeout = 24;
 
+  // If absent, allows safe requests by default.
+  oneof early_data {
+    // If true, allow `safe HTTP requests <https://www.rfc-editor.org/rfc/rfc7231#section-4.2.1>`_ to be sent on early data. Otherwise, do not use early data.
+    bool early_data_allows_safe_requests = 39;
+
+    // If specified, only the requests with :ref:`allowed methods <envoy_v3_api_field_config.route.v3.RouteAction.EarlyDataOptions.allowed_methods>` can use early data.
+    EarlyDataOptions early_data_options = 40;
+  }
```

Comment:
> I'd like to confirm my understanding of how this works, as I'm a bit new to protos. This early_data field is optional, but if present contains either a bool early_data_allows_safe_requests or early_data_options. Is that right? If no early_data field is present, then early data is used for safe request. If it is present and early_data_allows_safe_requests is true, then early data is used for safe requests. If it is present and early_data_allows_safe_requests is false, then no early data is used? If it is present and early_data_options contains a list of methods, then methods matching this list will use early data. If it is present and early_data_options is empty then early data will not be used. Is that right?

Your description is basically correct, except that `early_data` is not a field in the proto, but either early_data_allows_safe_requests or early_data_options to be a field there. See https://developers.google.com/protocol-buffers/docs/proto3#oneof

> This sounds reasonable, though it did make me wonder if it might be simpler to have two members: bool allow_early_data EarlyDataOptions early_data_options
> 
> If allow_early_data is false, then early data is not used. Otherwise if early_data_options is omitted then early data is used for safe methods. Otherwise early data is only used for methods in that list. Would that be simpler?

I chose `oneof` to save a field in the config since, in theory, only early_data_allows_safe_requests or early_data_options needs to be present to achieve what we want to configure. And oneof is widely used in Envoy config and actually doesn't complicate the logic too much in control plane. But if we want this to be easier to understand, having both of them present in the config is fine. @markdroth for thoughts on this?

Reply:
Using a `oneof` seems like the appropriate option here, given that there's no use-case where we'd expect both fields to be set.

(As an aside, note that the `oneof` name itself does not appear in the protobuf generated API in C++, but that's not true in all languages.  In particular, in golang, the `oneof` name *does* appear in the protobuf generated API, which is why it would be considered a breaking change to move an existing field into or out of a `oneof`.  So once we add this field in a `oneof`, we won't be able to move it out if we decide later that it should not be a `oneof`; instead, we'd have to deprecate the existing field and add a new one.)